python/cudf/cudf/core/buffer/spill_manager.py

```diff
@@ -0,0 +1,363 @@
+# Copyright (c) 2022, NVIDIA CORPORATION.
+
+from __future__ import annotations
+
+import gc
+import io
+import os
+import threading
+import traceback
+import warnings
+import weakref
+from collections import defaultdict
+from dataclasses import dataclass
+from typing import Dict, List, MutableMapping, Optional, Tuple
+
+import rmm.mr
+
+from cudf.core.buffer.spillable_buffer import SpillableBuffer
+from cudf.utils.string import format_bytes
+
+
+def get_traceback() -> str:
+    with io.StringIO() as f:
+        traceback.print_stack(file=f)
+        f.seek(0)
+        return f.read()
+
+
+def get_rmm_memory_resource_stack(
+    mr: rmm.mr.DeviceMemoryResource,
+) -> List[rmm.mr.DeviceMemoryResource]:
+    if hasattr(mr, "upstream_mr"):
+        return [mr] + get_rmm_memory_resource_stack(mr.upstream_mr)
+    return [mr]
```

Comment:
Rather than returning a list and requiring recursion, I would implement this as a generator
```suggestion
def get_rmm_memory_resource_stack(
    mr: rmm.mr.DeviceMemoryResource,
) -> List[rmm.mr.DeviceMemoryResource]:
    while True:
        yield mr
        if not hasattr(mr, "upstream_mr"):
            break
```

Reply:
Since this is not performance sensitive, I guess it is a matter of preference. I must admit I find the current implementation more elegant :)